### ISSUE TITLE
stability: OOM on blue shortly after startup

### ISSUE THREAD

**Issue by mberhault** [cockroachdb/cockroach#10458]
build sha: 81eaef6

`cockroach-blue-01` (`40.79.33.40`) OOMed shortly after chaos-restart (chaos was just restarted, it was the first node to be restarted in 3 hours or so).

timeline:
```
2016-11-04 15:00:01: kill by chaos monkey
2016-11-04 15:06:01: restart by chaos monkey
2016-11-04 15:08:40: kill by OOM killer
```

A goroutine grab at 15:08:16 shows 12k goroutines stuck on:
```
12339: select [0~2 minutes] [Created by grpc.(*Server).serveStreams.func1 @ server.go:423]
    storage             replica.go:852             (*Replica).redirectOnOrAcquireLease(*, #157774, *, *)
    storage             replica.go:1651            (*Replica).tryAddWriteCmd(*, #157774, *, *, *, #146, *, *, 0, *, ...)
    storage             replica.go:1610            (*Replica).addWriteCmd(*, #157774, *, *, *, #146, *, *, 0, *, ...)
    storage             replica.go:1104            (*Replica).Send(*, #157774, *, *, *, #146, *, *, 0, *, ...)
    storage             store.go:2332              (*Store).Send(#1320, #157774, *, *, *, #146, *, *, 0, *, ...)
    storage             stores.go:187              (*Stores).Send(#405, #157774, *, 0, 0, #146, *, *, 0, *, ...)
    server              node.go:844                (*Node).batchInternal.func1(0, 0)
    stop                stopper.go:245             (*Stopper).RunTaskWithErr(#306, *, 0, 0)
    server              node.go:855                (*Node).batchInternal(#618, #157774, *, *, *, *, *)
    server              node.go:875                (*Node).Batch(#618, #157774, *, *, *, *, *)
    roachpb             api.pb.go:1521             _Internal_Batch_Handler(#39, #618, #157774, *, *, 0, 0, 0, *, *)
    grpc                server.go:611              (*Server).processUnaryRPC(#831, #104, *, *, #833, #78, 0, 0, 0)
    grpc                server.go:769              (*Server).handleStream(#831, #104, *, *, 0)
    grpc                server.go:422              (*Server).serveStreams.func1.1(*, #831, #104, *, *)
```

Full goroutine dump (run through panicparse):
[stacks.txt](https://github.com/cockroachdb/cockroach/files/571936/stacks.txt)

Heap profile from 15:08:32 shows:
```
Showing top 10 nodes out of 50 (cum >= 25.28MB)
      flat  flat%   sum%        cum   cum%
 2295.54MB 61.47% 61.47%  2295.54MB 61.47%  github.com/coreos/etcd/raft/raftpb.(*Entry).Unmarshal
  906.90MB 24.28% 85.75%   907.40MB 24.30%  google.golang.org/grpc.(*parser).recvMsg
  282.64MB  7.57% 93.32%  2577.69MB 69.02%  github.com/coreos/etcd/raft/raftpb.(*Message).Unmarshal
   29.75MB   0.8% 94.12%    39.76MB  1.06%  github.com/cockroachdb/cockroach/pkg/roachpb.(*BatchRequest).Unmarshal
   22.33MB   0.6% 94.72%    22.33MB   0.6%  github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatchBuilder).grow
   20.97MB  0.56% 95.28%  2598.65MB 69.59%  github.com/cockroachdb/cockroach/pkg/storage.(*RaftMessageRequestBatch).Unmarshal
   19.21MB  0.51% 95.79%    19.21MB  0.51%  google.golang.org/grpc/transport.(*http2Server).handleData
    1.50MB  0.04% 95.83%   100.50MB  2.69%  google.golang.org/grpc.(*Server).processUnaryRPC
       1MB 0.027% 95.86%       93MB  2.49%  github.com/cockroachdb/cockroach/pkg/roachpb._Internal_Batch_Handler
    0.50MB 0.013% 95.87%    25.28MB  0.68%  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).tryAddWriteCmd
```

Full graph:
[heap.pdf](https://github.com/cockroachdb/cockroach/files/571945/heap.pdf)

Logs are saved on `40.79.33.40:logs.10458`

**Comment by mberhault** [comment]
May or may not be related to #10427.

**Comment by petermattis** [comment]
The goroutine spike seems to be happening again.

**Comment by mberhault** [comment]
yup, same node, same timing. memory is not spiking as quickly though

**Comment by petermattis** [comment]
@bdarnell `intentResolver` sets a 3s timeout on operations, but won't that have a bad interaction with the 9s lease expiration that we might have to wait for when a node dies abruptly?

**Comment by bdarnell** [comment]
The stack traces are pretty different from #10427. 

When a node is down, attempts to resolve intents on it will fail, but I don't see how the two timeouts are connected (it should be failing fast instead of waiting out the 3s timeout). Even if it were waiting out the 3s, we might end up with 3x as many queued ops as we need but it doesn't seem like a major problem.

**Comment by bdarnell** [comment]
OK, looking at the full goroutine dumps I see a few goroutines actively working in the command queue, so this could be an early sign of #10427 before it hits the critical tipping point. Only 13 goroutines blocked on the mutex in beginCmds, though.

```
1: runnable [Created by grpc.(*Server).serveStreams.func1 @ server.go:423]
    interval            interval.go:98             Comparable.Compare(#89529, 0x1, 0x1, #104670, 0, 0x1, #157366)
    interval            interval.go:64             Range.OverlapExclusive(#89529, 0, 0x1, #89529, 0x1, 0x1, #104670, 0, 0x1, #104670, ...)
    interval            interval.go:685            (*Node).doMatch(#137377, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104670)
    interval            interval.go:692            (*Node).doMatch(#137376, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104669)
    interval            interval.go:680            (*Node).doMatch(#137378, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104671)
    interval            interval.go:680            (*Node).doMatch(#137379, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104672)
    interval            interval.go:680            (*Node).doMatch(#137380, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104673)
    interval            interval.go:692            (*Node).doMatch(#137375, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104668)
    interval            interval.go:692            (*Node).doMatch(#137374, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104667)
    interval            interval.go:680            (*Node).doMatch(#137381, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104674)
    interval            interval.go:680            (*Node).doMatch(#137382, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #104675)
    interval            interval.go:680            (*Node).doMatch(#108642, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #137117)
    interval            interval.go:680            (*Node).doMatch(#77526, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #119867)
    interval            interval.go:680            (*Node).doMatch(#114754, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #115062)
    interval            interval.go:692            (*Node).doMatch(#105893, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, #54, #115062)
    interval            interval.go:673            (*Tree).DoMatching(#694, #157367, #89529, 0, 0x1, #89529, 0x1, 0x1, 0x1)
    storage             command_queue.go:338       (*CommandQueue).getOverlaps(#694, #89529, 0, 0x1, #89529, 0x1, 0x1, #157406, 0x1c23, 0x2400)
    storage             command_queue.go:177       (*CommandQueue).getWait(#694, #17245, #144727, 0x15a, 0x15a, 0, 0, 0)
    storage             replica.go:1222            (*Replica).beginCmds(#1515, #157774, #15450, #43317, #157368, #89528, 0)
    storage             replica.go:1635            (*Replica).tryAddWriteCmd(#1515, #157774, #15450, #171757, 0, #146, 0x1c, 0xc0c, 0, 0, ...)
    storage             replica.go:1610            (*Replica).addWriteCmd(#1515, #157774, #15450, #171757, 0, #146, 0x1c, 0xc0c, 0, 0, ...)
    storage             replica.go:1104            (*Replica).Send(#1515, #157774, #15450, #171757, 0, #146, 0x1c, 0xc0c, 0, 0, ...)
    storage             store.go:2332              (*Store).Send(#1320, #157774, #15449, #171757, 0, #146, 0x1c, 0xc0c, 0, 0, ...)
    storage             stores.go:187              (*Stores).Send(#405, #157774, #15448, 0, 0, #146, 0x1c, 0xc0c, 0, 0, ...)
    server              node.go:844                (*Node).batchInternal.func1(0, 0)
    stop                stopper.go:245             (*Stopper).RunTaskWithErr(#306, #157369, 0, 0)
    server              node.go:855                (*Node).batchInternal(#618, #157774, #15447, #43313, #15447, #104770, 0x603fa7)
    server              node.go:875                (*Node).Batch(#618, #157774, #15447, #43313, 0, 0x81e, 0x81e)
    roachpb             api.pb.go:1521             _Internal_Batch_Handler(#39, #618, #157774, #82868, #157336, 0, 0, 0, 0, 0)
    grpc                server.go:611              (*Server).processUnaryRPC(#831, #104, #2646, #6863, #833, #78, 0, 0, 0)
    grpc                server.go:769              (*Server).handleStream(#831, #104, #2646, #6863, 0)
    grpc                server.go:422              (*Server).serveStreams.func1.1(#1328, #831, #104, #2646, #6863)
```

The memory is mostly raft log entries, though, and the heap profile indicates that they're being read from the network and not the disk. These messages should be written to disk and released from memory on the next raft.Ready cycle. There are no idle raft workers, although all their stack traces look normal and I don't think any of them are getting wedged somewhere. 

If we were simply unable to keep up with the load downstream of raft, then an accumulation of raftpb.Entires in memory would be one symptom we'd see. (maybe just after restart we've got a cold cache and have a harder time keeping up?) I think the fix for that would be backpressure in `beginCmds` based on the size of the uncommitted tail of the raft log.

**Comment by mberhault** [comment]
saving logs from delta node `104.196.115.151` on `~/logs.10458`.
OOMed a few hours ago.

latest heap profile:
```
2758.69MB of 3289.41MB total (83.87%)
Dropped 1846 nodes (cum <= 16.45MB)
Showing top 10 nodes out of 119 (cum >= 38.49MB)
      flat  flat%   sum%        cum   cum%
 1005.28MB 30.56% 30.56%  1005.28MB 30.56%  github.com/coreos/etcd/raft/raftpb.(*Entry).Unmarshal
  593.78MB 18.05% 48.61%   649.79MB 19.75%  github.com/cockroachdb/cockroach/pkg/roachpb.(*BatchRequest).Unmarshal
  506.28MB 15.39% 64.00%   506.28MB 15.39%  github.com/cockroachdb/cockroach/pkg/storage.(*SnapshotRequest).Unmarshal
  134.15MB  4.08% 68.08%   134.15MB  4.08%  github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatchBuilder).grow
  133.40MB  4.06% 72.14%  1277.98MB 38.85%  github.com/cockroachdb/cockroach/pkg/storage.(*RaftMessageRequestBatch).Unmarshal
  130.28MB  3.96% 76.10%  1134.56MB 34.49%  github.com/coreos/etcd/raft/raftpb.(*Message).Unmarshal
  104.45MB  3.18% 79.27%   104.95MB  3.19%  google.golang.org/grpc.(*parser).recvMsg
   68.14MB  2.07% 81.34%    69.64MB  2.12%  google.golang.org/grpc/transport.(*http2Server).handleData
   44.44MB  1.35% 82.70%    46.95MB  1.43%  github.com/cockroachdb/cockroach/pkg/storage.(*Store).uncoalesceBeats
   38.49MB  1.17% 83.87%    38.49MB  1.17%  github.com/coreos/etcd/raft.(*raft).send
```

Full stack trace (through `panicparse --aggressive`):
[stack.txt](https://github.com/cockroachdb/cockroach/files/580216/stack.txt)

**Comment by petermattis** [comment]
This is curious:

```
I161121 17:08:38.219851 134 server/status/runtime.go:228  runtime stats: 15 GiB RSS, 3356 goroutines, 8.4 GiB/971 MiB/10 GiB GO alloc/idle/total, 4.0 GiB/5.1 GiB CGO alloc/total, 120429.66cgo/sec, 5.58/0.79 %(u/s)time, 0.00 %gc (0x)
I161121 17:15:48.717612 134 server/status/runtime.go:228  runtime stats: 15 GiB RSS, 3592 goroutines, 8.8 GiB/508 MiB/10 GiB GO alloc/idle/total, 4.0 GiB/5.1 GiB CGO alloc/total, 58553.89cgo/sec, 4.41/0.41 %(u/s)time, 0.00 %gc (1x)
I161121 17:15:48.760388 134 server/status/runtime.go:228  runtime stats: 14 GiB RSS, 3699 goroutines, 5.5 GiB/2.3 GiB/9.5 GiB GO alloc/idle/total, 4.0 GiB/5.1 GiB CGO alloc/total, 31.79cgo/sec, 0.06/0.04 %(u/s)time, 0.00 %gc (1x)
I161121 17:18:19.912636 134 server/status/runtime.go:228  runtime stats: 14 GiB RSS, 21733 goroutines, 7.2 GiB/1.9 GiB/10 GiB GO alloc/idle/total, 4.1 GiB/5.2 GiB CGO alloc/total, 14618.46cgo/sec, 0.28/0.06 %(u/s)time, 0.00 %gc (0x)
```

This was grepping for `runtime.go:228`. These log messages are supposed to be printed every 10 sec and elsewhere in the log they are. But check out those gaps. The first one is 7min and the next is 3min. 

In that first time period I see a ton of long lock hold time warnings:

```
W161121 17:15:48.667496 14389857 storage/replica.go:591  [n2,s3,r25/2:/Table/51/1/{689175…-703684…}] cmdQMu: mutex held by github.com/cockroachdb/cockroach/pkg/storage.(*endCmds).done for 16.046458718s (>500ms):
goroutine 14389857 [running]:
```

This one is particularly interesting because the command queue is only held during memory operations. So either something is massively horked with the process or the command queue was in some super inefficient state.

**Comment by petermattis** [comment]
We saw this again on `delta` while running 9c1531a2f67aadf0e1d5e0b8d96fa5861af300c4. The immediate symptom was:

```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
```

This was preceded by a lot of messages that looked like:

```
I161121 17:33:26.377043 112 util/stop/stopper.go:294  stopper throttling task from storage/intent_resolver.go:437 due to semaphore
```

And when I saw a lot, I mean a lot: 252645 occurrences. That was over the lifetime of the process (which was a little more than an hour).

Looking at stack traces, all 100 intent resolver tasks are blocked. About half of them are blocked trying to run the GC operation after an `EndTranaction` request. The other half are blocked earlier in the `intentResolver.resolveIntents` code. Looking at the raw stacks, it appears that they are all blocked on the same replica (`0xc5bb753000`). I'm not sure how to determine which replica that is, though.
